File: ruby/anagram/anagram.rb
class Anagram
  attr_reader :input

  def initialize(raw_input)
    @input = raw_input.downcase
  end

  def match(words)
    target = alphabetize(input)
    words.select { |word| anagram?(word, target) }
  end

  private

  def alphabetize(word)
    word.downcase.split('').sort.join
  end

  def anagram?(word, target)
    alphabetic_match?(word, target) && distinct_from_input?(word)
  end

  def alphabetic_match?(word, target)
    alphabetize(word) == target
  end

  def distinct_from_input?(word)
    word.downcase != input
  end
end

module BookKeeping
  VERSION = 2
end

File: ecmascript/anagram/babel-output/anagram.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var Anagram = (function () {
  function Anagram(input) {
    _classCallCheck(this, Anagram);

    this.input = input;
  }

  _createClass(Anagram, [{
    key: 'matches',
    value: function matches() {
      var _Array$prototype,
          _this = this;

      var words = (_Array$prototype = Array.prototype).concat.apply(_Array$prototype, arguments);
      return words.filter(function (word) {
        return _this.isAnagram(word);
      });
    }
  }, {
    key: 'alphabetize',
    value: function alphabetize(text) {
      return text.toLowerCase().split('').sort().join('');
    }
  }, {
    key: 'isAnagram',
    value: function isAnagram(word) {
      return this.isDistinctFromInput(word) && this.isAlphabeticMatch(word);
    }
  }, {
    key: 'isAlphabeticMatch',
    value: function isAlphabeticMatch(word) {
      return this.alphabetize(word) === this.alphabetize(this.input);
    }
  }, {
    key: 'isDistinctFromInput',
    value: function isDistinctFromInput(word) {
      return word.toLowerCase() !== this.input.toLowerCase();
    }
  }]);

  return Anagram;
})();

exports['default'] = Anagram;
module.exports = exports['default'];